### DISCUSSION TITLE
Multiple keyword search

### DISCUSSION THREAD

**Discussion by nyarly** [rust-lang/crates.io#4337] [Feature Requests]
I've helped take over maintenance of the Gotham web framework, and discussion started around how to package middleware for it.

What I'd like to suggest is that they be shipped as crates with the keywords `gotham` and `middleware`, since both would make sense, and then the Gotham docs could link to a page on crates.io that would list available middleware.

However, I can't seem to combine keywords in the crates.io/keywords/xxx route. It would be fab if e.g. I could link to crates.io/keywords/gotham+middleware and have that Just Work.

**Comment by pwoolcoc** [comment]
@nyarly do you see the `crates.io/keywords/gotham+middleware` idea as being more of an "AND" operation, or an "OR" operation? That is, if that URL existed would you expect to see crates that have the `gotham` keyword OR the `middleware` keyword, or crates that have both the `gotham` keyword AND the `middleware` keyword?

**Comment by nyarly** [comment]
I'd like a URL for the intersection: both `gotham` and `middleware`. Whether that's `+` or what I'm indifferent about.

**Comment by erlend-sh** [comment]
An AND operation for URLs would be immensely useful for us as well at 
http://arewegameyet.com/

The Rust gamedev ecosystem is growing fast and it’s becoming untenable to manually add every new library to our handcrafted list.

What is the gist of required work for this feature @pwoolcoc? We might have some people in our midsts who’d be able to contribute. I see some work has already gone into #1543 but I’m a bit confused as to what exactly it does and why there’s talk of `search`. Practically speaking what’s needed isn’t a search but just a filter.

**Comment by carols10cents** [comment]
@erlend-sh We do all queries, whether they use a free-form search term or a filter, through the API endpoint named `search`. I'm going to take a close look at #1543 today, there are some merge conflicts and sql query fixes to make, it looks like.

And as I understand it, that PR only makes it possible to construct URLs manually that do an AND keyword operation-- would that be sufficient for arewegameyet's purposes, or would you want a UI in crates.io with, say, checkboxes? Or even just displaying what the search results are filtering on....

I don't think we need a UI to merge the ability to manually construct queries, but it'd be great to have one eventually.

**Comment by erlend-sh** [comment]
A filter UI akin to GitHub issues or similar would of course be great but we can definitely get by fine without it.

> And as I understand it, that PR only makes it possible to construct URLs manually that do an AND keyword operation-- would that be sufficient for arewegameyet's purposes

That sounds sufficient, yes. Pretty sure we want the exact same thing as @nyarly, which is a link like `crates.io/keywords/gotham+middleware` which shows any crate that contains **both** of those keywords.

**Comment by nyarly** [comment]
This got converted (automatically?) to a discussion. I still think this would be a valuable and clearly defined feature. "Clearly defined" may only be true for me and @erlend-sh though :)

**Comment by nyarly** [comment]
Actually: the PR referred to above seems to solve the problem for us. Thanks, @carols10cents !
